wgpu/backends/rs.py

```diff
@@ -1535,76 +1533,87 @@ def callback(status_, user_data_p):
         self._map_state = enums.BufferMapState.pending
         # H: void f(WGPUBuffer buffer, WGPUMapModeFlags mode, size_t offset, size_t size, WGPUBufferMapCallback callback, void * userdata)
         libf.wgpuBufferMapAsync(
-            self._internal, lib.WGPUMapMode_Read, 0, size, callback, ffi.NULL
+            self._internal, map_mode, offset, size, callback, ffi.NULL
         )
 
         # Let it do some cycles
-        # H: void f(WGPUInstance instance)
-        # libf.wgpuInstanceProcessEvents(get_wgpu_instance())
         # H: bool f(WGPUDevice device, bool wait, WGPUWrappedSubmissionIndex const * wrappedSubmissionIndex)
         libf.wgpuDevicePoll(self._device._internal, True, ffi.NULL)
 
         if status != 0:  # no-cover
-            raise RuntimeError(f"Could not read buffer data ({status}).")
+            raise RuntimeError(f"Could not map buffer ({status}).")
         self._map_state = enums.BufferMapState.mapped
+        self._mapped_range = offset, offset + size
 
-        # Copy data
+    async def map_async(self, mode, offset=0, size=None):
+        return self.map()  # for now
+
+    def unmap(self):
+        # Can we even unmap?
+        if self._map_state != enums.BufferMapState.mapped:
+            raise RuntimeError("Can only unmap a buffer if its currently mapped.")
+        # H: void f(WGPUBuffer buffer)
+        libf.wgpuBufferUnmap(self._internal)
+        self._map_state = enums.BufferMapState.unmapped
+        self._mapped_range = None
+
+    def read_mapped(self, offset=0, size=None):
+        # Can we even read?
+        if self._map_state != enums.BufferMapState.mapped:
+            raise RuntimeError("Can only read from a buffer if its mapped.")
+
+        # Check offset and size
+        offset, size = self._check_range(offset, size)
+        if offset < self._mapped_range[0] or (offset + size) > self._mapped_range[1]:
+            raise ValueError(
+                "The range for buffer reading is not contained in the currently mapped range."
+            )
+
+        # Get mapped memoryview.
         # H: void * f(WGPUBuffer buffer, size_t offset, size_t size)
         src_ptr = libf.wgpuBufferGetMappedRange(self._internal, 0, size)
         src_address = int(ffi.cast("intptr_t", src_ptr))
         src_m = get_memoryview_from_address(src_address, size)
+
+        # Copy the data. The memoryview created above becomes invalid when the buffer
+        # is unmapped, so we don't want to pass that memory to the user.
+        data = memoryview((ctypes.c_uint8 * size)()).cast("B")
```

Comment:
you could add a kwarg (I'm not sure if that would conflict with the IDL) like `copy=True` to allow users to opt in to the dangerous behaviour?

Reply:
That sounds reasonable! At least in that case its a very deliberate choice.

Reply:
Now we check all 4 of the boxes 🚀